engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/typer/TypeMethodReference.scala

```diff
@@ -27,39 +25,64 @@ class TypeMethodReference(methodReference: MethodReference, currentResults: List
-          case methodInfoes => typeFromMethodInfoes(methodInfoes)
-        }
+  private def typeFromClazzDefinitions(clazzDefinitions: List[ClazzDefinition]): Either[String, TypingResult] = {
+    val validatedType = for {
+      nonEmptyClassDefinitions <- validateClassDefinitionsNonEmpty(clazzDefinitions).right
+      nonEmptyMethods <- validateMethodsNonEmpty(nonEmptyClassDefinitions).right
+      methodsWithMatchingArity <- validateMethodArity(nonEmptyMethods).right
+      returnTypesForMatchingParams <- validateMethodParameterTypes(methodsWithMatchingArity).right
+    } yield Typed(returnTypesForMatchingParams.toSet)
+
+    validatedType match {
+      case Left(None) => Right(Unknown) // we use Left(None) to indicate situation when we want to skip further validations because of lack of some knowledge
+      case Left(Some(message)) => Left(message)
+      case Right(returnType) => Right(returnType)
+    }
+  }
+
+  private def validateClassDefinitionsNonEmpty(clazzDefinitions: List[ClazzDefinition]): Either[Option[String], List[ClazzDefinition]] =
+    if (clazzDefinitions.isEmpty) Left(None) else Right(clazzDefinitions)
+
+  private def validateMethodsNonEmpty(clazzDefinitions: List[ClazzDefinition]): Either[Option[String], List[MethodInfo]] = {
+    def displayableType = clazzDefinitions.map(k => k.clazzName).map(_.display).mkString(", ")
+    def isClass = clazzDefinitions.map(k => k.clazzName).exists(_.canBeSubclassOf(Typed[Class[_]]))
+    clazzDefinitions.flatMap(_.methods.get(methodName)) match {
+      //Static method can be invoked - we cannot find them ATM
+      case Nil if isClass => Left(None)
+      case Nil => Left(Some(s"Unknown method '$methodName' in $displayableType"))
+      case methodInfoes => Right(methodInfoes)
     }
+  }
 
-  //TODO: we check only arity, but don't check if any of overloaded methods has matching signature
-  //this will lead to erros if we have different return types for different signatures!
-  private def typeFromMethodInfoes(methodInfoes: List[MethodInfo]): Either[String, TypingResult] =
-    methodInfoes.filter(_.parameters.size <= paramsCount) match {
+  private def validateMethodArity(methodInfoes: List[MethodInfo]): Either[Option[String], List[MethodInfo]] = {
+    methodInfoes.filter(_.parameters.size <= calledParams.size) match {
       case Nil =>
-        Left(s"Invalid arity for '${methodReference.getName}'")
+        Left(Some(s"Invalid arity for '$methodName'"))
       case nonEmpty =>
-        val returnTypes = nonEmpty.map(_.refClazz)
-        val typingResult = Typed(returnTypes.toSet)
-        Right(typingResult)
+        Right(nonEmpty)
     }
+  }
 
+  private def validateMethodParameterTypes(methodInfoes: List[MethodInfo]): Either[Option[String], List[TypingResult]] = {
+    val returnTypesForMatchingMethods = methodInfoes.flatMap { m =>
+      val allMatching = m.parameters.map(_.refClazz).zip(calledParams).forall {
```

Comment:
this zip can omit some subtle edge cases when we have not same arity, but same types in common prefix?

Reply:
Indeed, there is hidden logic that it based on result of `validateMethodArity`. I can merge both methods so it will be more visible that we can't call methods passing less arguments than it is required

Reply:
I've removed `validateMethodArity` and added checking of sizes here.